pkg/translator/pprof/pprof_to_profiles.go

```diff
@@ -0,0 +1,639 @@
+	// unit is associated to this Attribute.
+	noAttrUnit = int32(-1)
+)
+
+// attr is a helper struct to build pprofile.ProfilesDictionary.attribute_table.
+type attr struct {
+	keyStrIdx  int32
+	value      any
+	unitStrIdx int32
+}
+
+// fn is a helper struct to build pprofile.ProfilesDictionary.function_table.
+type fn struct {
+	name       string
+	systemName string
+	fileName   string
+	startLine  int64
+}
+
+// mm is a helper struct to build pprofile.ProfilesDictionary.mapping_table.
+type mm struct {
+	memoryStart    uint64
+	memoryLimit    uint64
+	fileOffset     uint64
+	filenameStrIdx int32
+	attrIdxs       string // List of consecutive increasing indices separated by a semicolon.
+}
+
+// loc is a helper struct to build pprofile.ProfilesDictionary.location_table.
+type loc struct {
+	mappingIdx int32
+	address    uint64
+	lines      string // String representation of a sorted list of lines.
+	attrIdxs   string // List of consecutive increasing indices separated by a semicolon.
+}
+
+// lookupTables is a helper struct around pprofile.ProfilesDictionary.
+type lookupTables struct {
+	mappingTable        map[mm]int32
+	lastMappingTableIdx int32
+
+	locationTable        map[loc]int32
+	lastLocationTableIdx int32
+
+	functionTable        map[fn]int32
+	lastFunctionTableIdx int32
+
+	// The concept of profiles.Link does not exit in pprof.
+	// Therefore this helper table is skipped.
+
+	stringTable        map[string]int32
+	lastStringTableIdx int32
+
+	attributeTable        map[attr]int32
+	lastAttributeTableIdx int32
+
+	stackTable        map[string]int32
+	lastStackTableIdx int32
+}
+
+func convertPprofToPprofile(src *profile.Profile) (*pprofile.Profiles, error) {
+	if err := src.CheckValid(); err != nil {
+		return nil, fmt.Errorf("%w: %w", err, errInvalPprof)
+	}
+	dst := pprofile.NewProfiles()
+
+	// Initialize remaining lookup tables of pprofile.ProfilesDictionary in initLookupTables.
+	lts := initLookupTables()
+
+	// Add envelope messages
+	rp := dst.ResourceProfiles().AppendEmpty()
+	rp.SetSchemaUrl(semconv.SchemaURL)
+
+	sp := rp.ScopeProfiles().AppendEmpty()
+	sp.SetSchemaUrl(semconv.SchemaURL)
+
+	// Use a dedicated pprofile.Profile for each sample type.
+	for stIdx, st := range src.SampleType {
+		p := sp.Profiles().AppendEmpty()
+
+		// pprof.Profile.sample_type
+		p.SampleType().SetTypeStrindex(lts.getIdxForString(st.Type))
+		p.SampleType().SetUnitStrindex(lts.getIdxForString(st.Unit))
+
+		// pprof.Profile.sample
+		for _, sample := range src.Sample {
+			s := p.Samples().AppendEmpty()
+
+			// pprof.Sample.location_id
+			stackIdx := lts.getIdxForStack(sample.Location)
+			s.SetStackIndex(stackIdx)
+
+			// pprof.Sample.value
+			s.Values().Append(sample.Value[stIdx])
+
+			// pprof.Sample.label - this field is split into string and numeric labels.
+			for lk, lv := range sample.Label {
+				if len(lv) != 1 {
+					return nil, fmt.Errorf("invalid length of string label value %d: %w",
+						len(lv), errInvalPprof)
```

Comment:
Technically having multiple label values for the same key is valid pprof. I don't think it's critical that we handle it here, but the error message should probably read more like: "labels with multiple values are not supported" rather than claiming that the pprof itself is invalid?